nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java

```diff
@@ -0,0 +1,431 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.nessie;
+
+import com.dremio.nessie.api.TreeApi;
+import com.dremio.nessie.client.NessieClient;
+import com.dremio.nessie.error.NessieConflictException;
+import com.dremio.nessie.error.NessieNotFoundException;
+import com.dremio.nessie.model.Contents;
+import com.dremio.nessie.model.ContentsKey;
+import com.dremio.nessie.model.EntriesResponse;
+import com.dremio.nessie.model.IcebergTable;
+import com.dremio.nessie.model.ImmutableDelete;
+import com.dremio.nessie.model.ImmutableOperations;
+import com.dremio.nessie.model.ImmutablePut;
+import com.dremio.nessie.model.Operations;
+import com.dremio.nessie.model.Reference;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.BaseMetastoreCatalog;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+/**
+ * Nessie implementation of Iceberg Catalog.
+ *
+ * <p>
+ *   A note on namespaces: Nessie namespaces are implicit and do not need to be explicitly created or deleted.
+ *   The create and delete namespace methods are no-ops for the NessieCatalog. One can still list namespaces that have
+ *   objects stored in them to assist with namespace-centric catalog exploration.
+ * </p>
+ */
+public class NessieCatalog extends BaseMetastoreCatalog implements AutoCloseable, SupportsNamespaces, Configurable {
+
+  private static final Joiner SLASH = Joiner.on("/");
+  public static final String NESSIE_WAREHOUSE_DIR = "nessie.warehouse.dir";
+  private NessieClient client;
+  private String warehouseLocation;
+  private Configuration config;
+  private UpdateableReference reference;
+  private String name;
+
+  /**
+   * Try to avoid passing parameters via hadoop config. Dynamic catalog expects Map instead
+   *
+   * todo replace with #1640 style constructor
```

Comment:
#1640 is in. It uses a no-arg constructor and adds an `initialize(String name, Map<String, String> config)` method to initialize and configure the catalog. I think you should be able to update this now.

I'm hoping that this removes the need to make Spark and Flink depend on the new Nessie and Glue modules. We should make sure we have a test suite we can include here that uses Flink and Spark.

Reply:
This is fixed now, apologies. I fixed the constructor but didn't remove the comment.

I agree that the Catalog portion of Spark3 should work fine now w/o explicitly adding Nessie (or Glue etc). I believe we still need to update the `IcebergSource` to handle custom (Iceberg) catalogs right?

Is the intention to add the new catalogs to the Iceberg shaded jar?

Reply:
> Is the intention to add the new catalogs to the Iceberg shaded jar?

I think it depends. If a catalog pulls in a ton of dependencies and requires updating a lot of the shaded Jar's documentation, then it comes at a high cost. On the other hand, if it uses existing bundled libraries or libraries that can be pulled from the Spark runtime, then it would be easier.

> I believe we still need to update the IcebergSource to handle custom (Iceberg) catalogs right?

Yes, we will need to come up with a way for IcebergSource to work with custom catalogs. Spark has a way for the source to return a catalog and identifier that is used instead of the source directly. That's a much better model, but the problem is that we don't necessarily know what the catalog should be. And, if we redirect to a catalog, we will need to also have a catalog that can load Hadoop tables from a URI. I think this is more of a follow-up.

Reply:
Which mechanism were you thinking for this? Was it `LookupCatalog`? That is rather scala-y but it works. I have something basic working along the lines of `LookupCatalog` but I have some concerns about it. Shall I post a PR as a straw man?